phpunit/html/wp-html-tag-processor-test.php

```diff
@@ -235,6 +246,68 @@ public function test_set_attribute_on_a_non_existing_tag_does_not_change_the_mar
 		);
 	}
 
+	/**
+	 * Passing a double quote inside of an attribute values could lead to an XSS attack as follows:
+	 *
+	 * <code>
+	 *     $p = new WP_HTML_Tag_Processor( '<div class="header"></div>' );
+	 *     $p->next_tag();
+	 *     $p->set_attribute('class', '" onclick="alert');
+	 *     echo $p;
+	 *     // <div class="" onclick="alert"></div>
+	 * </code>
+	 *
+	 * To prevent it, `set_attribute` calls `esc_attr()` on its given values.
+	 *
+	 * <code>
+	 *    <div class="&quot; onclick=&quot;alert"></div>
+	 * </code>
+	 *
+	 * @ticket 56299
+	 *
+	 * @dataProvider data_set_attribute_escapable_values
+	 * @covers set_attribute
+	 */
+	public function test_set_attribute_prevents_xss( $attribute_value ) {
+		$p = new WP_HTML_Tag_Processor( '<div></div>' );
+		$p->next_tag();
+		$p->set_attribute( 'test', $attribute_value );
+
+		/*
+		 * Testing the escaping is hard using tools that properly parse
+		 * HTML because they might interpret the escaped values. It's hard
+		 * with tools that don't understand HTML because they might get
+		 * confused by improperly-escaped values.
+		 *
+		 * For this test, since we control the input HTML we're going to
+		 * do what looks like the opposite of what we want to be doing with
+		 * this library but are only doing so because we have full control
+		 * over the content and because we want to look at the raw values.
+		 */
+		$match = null;
+		preg_match( '~^<div test=(.*)></div>$~', (string) $p, $match );
+		list( $full_match, $actual_value ) = $match;
+
+		$this->assertEquals( $actual_value, '"' . esc_attr( $attribute_value ) . '"' );
+	}
+
+	/**
+	 * Data provider with HTML attribute values that might need escaping.
+	 */
+	public function data_set_attribute_escapable_values() {
+		return array(
+			[ '"' ],
+			[ '&quot;' ],
+			[ '&' ],
+			[ '&amp;' ],
+			[ '&euro;' ],
+			[ "'" ],
+			[ '<>' ],
+			[ '&quot";' ],
+			[ '" onclick="alert(\'1\');"><span onclick=""></span><script>alert("1")</script>' ],
```

Comment:
I removed the tests that asserted we always end up with double quotes. that assertion is implicit in the `preg_match`.

is that relevant or necessary here to assert that? if so, I wonder if that would be best left for a different test just for that behavior.

Reply:
I was hesitant at first, but after staring at it for a while I think it's a much nicer way of approaching this test. Thank you 🙇

Reply:
I added another test file in 8cf66b0661. I understand your point in another discussion that we don't need to re-assert the `esc_attr` function. My intention is to make sure something flashes red if WP_HTML_Tag_Procesor changes something in the implementation and no longer leads to the same result as a plain `esc_attr` would.

Reply:
> We could test that in another file that would be run with all WordPress functions loaded and then not run it locally.

ultimately the CI runs are what matters. in fact, if we're going to create a new test file and special instructions just to run with WordPress as in 8cf66b0 I'd rather just remove the special-casing to run the test locally.

the comment at the top of the file is nice, but would be redundant if we added it everywhere. if someone is like me and prefers to run this locally then I think they/I can temporarily modify the file. if someone wants to be able to test this project locally then we can hopefully catch _all_ files in one project to make the testing configuration simpler and more flexible.

Reply:
Followed up in https://github.com/WordPress/gutenberg/pull/44593